mlir/lib/Quantum/Transforms/DecomposeLoweringImpl.hpp

```diff
@@ -0,0 +1,424 @@
+/// The struct uses std::variant to ensure only one type is active at a time,
+/// preventing invalid states.
+///
+/// Example usage:
+///   QubitIndex dynamicIdx(operandValue);     // Runtime qubit index
+///   QubitIndex staticIdx(IntegerAttr::get(...)); // Compile-time constant
+///   QubitIndex invalidIdx;                   // Uninitialized state
+///
+///   if (dynamicIdx) {                        // Check if valid
+///     if (dynamicIdx.isValue()) {            // Check if runtime value
+///       Value idx = dynamicIdx.getValue();   // Get the Value
+///     }
+///   }
+class QubitIndex {
+  private:
+    // use monostate to represent the invalid index
+    std::variant<std::monostate, Value, IntegerAttr> index;
+    Value qreg;
+
+  public:
+    QubitIndex() : index(std::monostate()), qreg(nullptr) {}
+    QubitIndex(Value val, Value qreg) : index(val), qreg(qreg) {}
+    QubitIndex(IntegerAttr attr, Value qreg) : index(attr), qreg(qreg) {}
+
+    bool isValue() const { return std::holds_alternative<Value>(index); }
+    bool isAttr() const { return std::holds_alternative<IntegerAttr>(index); }
+    operator bool() const { return isValue() || isAttr(); }
+    Value getReg() const { return qreg; }
+    Value getValue() const { return isValue() ? std::get<Value>(index) : nullptr; }
+    IntegerAttr getAttr() const { return isAttr() ? std::get<IntegerAttr>(index) : nullptr; }
+};
+
+// The goal of this class is to analyze the signature of a custom operation to get the enough
+// information to prepare the call operands and results for replacing the op to calling the
+// decomposition function.
+class BaseSignatureAnalyzer {
+  protected:
+    bool isValid = true;
+
+    llvm::SmallVector<mlir::Value, 4> paramsStorage;
+
+    // Unified Signature Structure: All parameters, regardless of source (params or theta),
+    // are stored in a ValueRange for generalized processing.
+    struct Signature {
+        mlir::ValueRange params;
+        mlir::ValueRange inQubits;
+        mlir::ValueRange inCtrlQubits;
+        mlir::ValueRange inCtrlValues;
+        mlir::ValueRange outQubits;
+        mlir::ValueRange outCtrlQubits;
+
+        // Qreg mode specific information (assuming QubitIndex is defined)
+        llvm::SmallVector<QubitIndex> inWireIndices;
+        llvm::SmallVector<QubitIndex> inCtrlWireIndices;
+        llvm::SmallVector<QubitIndex> outQubitIndices;
+        llvm::SmallVector<QubitIndex> outCtrlQubitIndices;
+    } signature;
+
+    BaseSignatureAnalyzer(mlir::Operation *op, mlir::ValueRange params, mlir::ValueRange inQubits,
+                          mlir::ValueRange inCtrlQubits, mlir::ValueRange inCtrlValues,
+                          mlir::ValueRange outQubits, mlir::ValueRange outCtrlQubits,
+                          bool enableQregMode)
+        : paramsStorage(params.begin(), params.end()),
+          signature(Signature{.params = mlir::ValueRange(paramsStorage),
+                              .inQubits = inQubits,
```

Comment:
more clear/clean to use designated initializers.

Reply:
Thanks! Fixed [0a5ca61](https://github.com/PennyLaneAI/catalyst/pull/2160/commits/0a5ca61cf2c40950462989683874b93d33e440f6)